Update temperature cost function for symbiotic Nfix in FUN

wwieder opened the pull request:
### Description of changes
Updates temperature cost function of Nfix using lab measurements from [Bytnerowicz et al. (2022)](https://doi.org/10.1038/s41477-021-01090-x), see Supplementary Table 3.  

### Specific notes
Replaces the Houlton temperature cost function that's currently used in CNFUN.
- requires updated parameter file with min, optimum, and max temperatures for tropical and extra-tropical PFTs. `/glade/work/wwieder/modified_inputs/ctsm60_params_nfix.c241101.nc`

Updated temperature function (below in blue) compared to the default  temperature function from Houlton et al (orange) to calculate C costs of N fix in FUN (second figure).

![image](https://github.com/user-attachments/assets/a4752fd0-7c96-408e-a175-f7b3fb5b08cf)
![image](https://github.com/user-attachments/assets/d055f028-d486-4e85-814c-8f367efa6f4b)

**Contributors other than yourself, if any:**
Tom Bytnerowicz, @tbytnero, t.bytnerowicz@ufl.edu

**CTSM Issues Fixed (include github issue #):**
None

**Are answers expected to change (and if so in what way)?**
Reduces temperate N fix  and vegetation C pools.  Also modifies temporal changes over historical and SSP scenarios.

Diagnostics from CRU-JRA forced runs at f19 [posted here](https://webext.cgd.ucar.edu/I20TR/ctsm530_f19_nfix_hist2/lnd/ctsm530_f19_nfix_hist2_2004_2023-ctsm530_f19_PPE_hist_2004_2023/setsIndex.html)

Issue on [LMWG-dev linked here](https://github.com/NCAR/LMWG_dev/issues/72)
Discussion on Nfix started on [LMWG-dev](https://github.com/NCAR/LMWG_dev/discussions/73)
 
**Any User Interface Changes (namelist or namelist defaults changes)?**
TODO add namelist option for Bytnerowicz vs. Houlton Nfix cost function

**Does this create a need to change or add documentation? Did you do so?**
Needs to be added to documentation, not currently done

**Testing performed, if any:**
Ran scientific tests in point simulations (CLM5.1 code base) and global 2 degree spinup, historical and SSP2-4.5 case.

linniahawkins commented:
Given that the new temperature cost function will be included in the ctsm6 release, we will need to re-run the One-at-a-time parameter ensemble. To do so, we will need a tag that is greater than ctsm5.3.010 with this new cost function added.  

We will also need a parameter file that is compatible with ctsm5.3.010 (ctsm60_params.c241017.nc) and adds the three new parameters from this function. 

I'll also need a restart from the 2-degree spinup Will already ran.

wwieder commented:
@linniahawkins and @ekluzek I've: 
- Removed unnecessary usermods from this PR
- Updated to the latest dev tag, 012.
- Added new parameters for Tom's nfix function to the default parameter file `ctsm60_params.c241017.nc`.  This can be used / seen here `paramfile = '/glade/work/wwieder/modified_inputs/ctsm60_params.c241114.nc'`
- Created & built a case to test that things still work `/glade/work/wwieder/ctsm5.3.0/ctsm5.3.012_f19_nfix_test` (not submitted yet).
- Initial condition files for f19 are here 
```
finidat = '/glade/derecho/scratch/wwieder/archive/ctsm530_f19_nfix_pSASU/rest/0161-01-01-00000/ctsm530_f19_nfix_pSASU.clm2.r.0161-01-01-00000.nc'
```
----

Next steps would be to:
1. Erik creates a branch tag for Linnia
  - Will / @olyson checks 1850 survivability in the default f19_PPE case and the f19_nfix cases
3. Will / Erik adds a namelist option for the nfix_temperature_method = 1 (Houlton function from CLM5)
4. Bring PR onto b4b-dev
5. Answer changing tag to master, setting nfix_temperature_method =2 (Tom's function) 

Let me know what you think of this plan and what I've missed.

wwieder commented:
Doesn't seem like the NFIX changes make a big impact on arctic survival. Dead grids (max ELAI < 0.1) for PPE and NFix cases are on the left and center.  max LAIs from the PPE where Nfix case is dead are on the right.

![image](https://github.com/user-attachments/assets/2d664469-1535-44e3-ae4b-959eb9e3d889)

ekluzek commented:
Testing on Derecho and Izumi are as expected other than the following which fail, because they need update parameter files

ERS_Ld3.f10_f10_mg37.I2000Clm60Bgc.derecho_intel.clm-ciso_cwd_hr        (RUN)
ERS_Ld3.f10_f10_mg37.I2000Clm60Bgc.derecho_intel.clm-ciso_cwd_hr--clm-matrixcnOn        (RUN)
SMS_Lm1.f10_f10_mg37.I1850Clm60Bgc.derecho_intel.clm-clm60_monthly_matrixcn_soilCN30    (RUN)

ekluzek commented:
Branch tag made: branch_tags/BNF_v2.n01_ctsm5.3.012

wwieder commented on src/biogeochem/CNFUNMod.F90:
```diff
@@ -488,6 +487,8 @@ subroutine CNFUN(bounds,num_soilc, filter_soilc,num_soilp&
   integer   :: fixer                             ! 0 = non-fixer, 1
   ! =fixer 
+  !TODO, make namelist option
+  integer   :: nfix_method = 2                   ! 1 = Houlton, 2 = Bytnerowicz
```
@slevis-lmwg tagging you here.

slevis-lmwg replied:
@wwieder thanks. When I return to this in the morning, I'm tempted to update to the most recent ctsm tag, to avoid updating after the fact. Let me know if you have a reason to advise against that.

wwieder replied:
I think that's a good idea

slevis-lmwg replied:
Rethinking this idea because you suggested merging to b4b-dev. I will start by rebasing this to b4b-dev and go from there. Sorry if this sounds confusing. I will figure it out :-)

slevis-lmwg commented:
I have checked out this branch to work on the namelist flag:
`git describe` returns `branch_tags/BNF_v2.n01_ctsm5.3.012`

slevis-lmwg commented:
Testing code mods with
`./create_test ERP_D_P64x2_Ld3.f10_f10_mg37.I2000Clm50BgcCru.derecho_intel.clm-flexCN_FUN -c /glade/campaign/cgd/tss/ctsm_baselines/ctsm5.3.013`
**PASS** Houlton as the default: b4b identical to 013
**EXPECTED FAIL** Misspell Bitnerovich in user_nl_clm: graceful fail
**PASS** Bytnerowicz in user_nl_clm: expected fail with diff from baseline and (expected?) fail COMPARE_base_rest

slevis-lmwg commented:
@wwieder my code mods are pushed and ready for your review.

I could not assign you the review because you own the PR :-)

wwieder reviewed (commented):
Thanks for adding in the namelist controls, @slevis-lmwg.  I'm assuming it will be straightforward for CLM5 physics to default to Houlton and CLM6 physics to default to Tom's function once we have it turned on?

slevis-lmwg commented:
> Thanks for adding in the namelist controls, @slevis-lmwg. I'm assuming it will be straightforward for CLM5 physics to default to Houlton and CLM6 physics to default to Tom's function once we have it turned on?

Currently it always defaults to Houlton but, yes, later we can change namelist_defaults_ctsm.xml to
```
<nfix_method>Houlton</nfix_method>
<nfix_method phys="clm6_0">Bytnerowicz</nfix_method>
```

slevis-lmwg commented:
> Bytnerowicz in user_nl_clm: expected fail with diff from baseline and (expected?) fail COMPARE_base_rest

I want to follow up regarding the COMPARE_base_rest failure above by
1. rerunning the test to generate new baseline
2. rerun again comparing to the new baseline: do I get the same failure?

slevis-lmwg commented:
@wwieder I still get the COMPARE_base_rest failure in my follow-up test above. The Bytnerowicz method may need a new variable sent to restart to give b4b restarts. We can discuss and/or I could try to resolve this before we discuss.

slevis-lmwg commented:
@wwieder I may have been quick to conclude that there was a problem. I now see that my "case2run" in the test used "Houlton" even though the base case used "Bytnerowicz" so I need to revisit my testing approach.

slevis-lmwg commented:
Ok, I just had to temporarily hardwire Bytnerowicz as the default and the COMPARE_base_rest test passed, so we're good.

slevis-lmwg commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -544,10 +544,10 @@ attributes from the config_cache.xml file (with keys converted to upper-case).
+<paramfile phys="clm6_0" >lnd/clm2/paramdata/ctsm60_params.c241119.nc</paramfile>
+<paramfile phys="clm5_1" >lnd/clm2/paramdata/ctsm60_params.c241119.nc</paramfile>
+<paramfile phys="clm5_0" >lnd/clm2/paramdata/clm50_params.c241119.nc</paramfile>
+<paramfile phys="clm4_5" >lnd/clm2/paramdata/clm45_params.c241119.nc</paramfile>
```
Reminder to 
- [ ] ./rimport the new paramfile when we're ready to merge this

slevis-lmwg replied:
Tests on derecho
PASS ./build-namelist_test.pl

Tests on derecho and izumi
IN PROGRESS `./run_sys_tests -s aux_clm -c ctsm5.3.013 --skip-generate`

slevis-lmwg replied:
On izumi this command shows that aux_clm was OK:
./cs.status.fails | grep -v NLCOMP | grep -v PASS | grep -v 'EXPECTED FAILURE' | grep -v 'me baseline files were mis'

slevis-lmwg replied:
On derecho this command shows a few failures that I need to resolve:
./cs.status.fails | grep -v NLCOMP | grep -v PASS | grep -v 'EXPECTED FAILURE'
```
    FAIL ERP_D_Ld5.ne30pg3_t232.IHistClm51Sp.derecho_intel.clm-default BASELINE ctsm5.3.013: DIFF
    FAIL ERP_D_Ld9.ne30pg3_t232.I1850Clm51BgcCropG.derecho_intel.clm-clm51cam6LndTuningMode BASELINE ctsm5.3.013: DIFF
    FAIL ERS_Ld3.f10_f10_mg37.I2000Clm60Bgc.derecho_intel.clm-ciso_cwd_hr RUN time=42
    FAIL ERS_Ld3.f10_f10_mg37.I2000Clm60Bgc.derecho_intel.clm-ciso_cwd_hr--clm-matrixcnOn RUN time=39
    FAIL SMS_Lm1.f10_f10_mg37.I1850Clm60Bgc.derecho_intel.clm-clm60_monthly_matrixcn_soilCN30 RUN time=23
```

slevis-lmwg replied:
The last 3 are expected [as per Erik's post](https://github.com/ESCOMP/CTSM/pull/2869#issuecomment-2487759533).

slevis-lmwg replied:
The two DIFF failures show this diff in lnd_in
```
<  paramfile = '/glade/campaign/cesm/cesmdata/inputdata/lnd/clm2/paramdata/ctsm60_params.c241119.nc'
---
>  paramfile = '/glade/campaign/cesm/cesmdata/inputdata/lnd/clm2/paramdata/ctsm51_params.c241017.nc'
```
and these paramfiles differ more than just due to the newly introduced params. Also the second of the two DIFF failures has a Clm60 counterpart that passes and clm51 tests go away in ctsm5.3.014:
`ERP_D_Ld9.ne30pg3_t232.I1850Clm60BgcCropG.derecho_intel.clm-clm60cam7LndTuningModeLDust`

So I consider aux_clm on derecho OK.

slevis-lmwg replied:
@ekluzek
`./rimport -file lnd/clm2/paramdata/clm50_params.c241119.nc`
returns this error
`/glade/p/cesmdata/cseg/cseg/inputdata/lnd/clm2/paramdata/clm50_params.c241119.nc does not exist`
Can you tell whether it's due to file permissions or a different problem that I should report to somebody?

slevis-lmwg replied:
./rimport will have to wait for when the tool/repository works again. Jim Edwards filed a ticket with help@ucar.edu.

slevis-lmwg commented:
[This comment](https://github.com/ESCOMP/CTSM/pull/2882#issuecomment-2570030311) suggests that I will likely need to open an issue/PR to debug some aspect of the code that was introduced here.
